contrib/scalar/scalar.c

```diff
@@ -600,252 +758,84 @@ static int cmd_clone(int argc, const char **argv)
-	return res;
-}
-
-static int stage_file(const char *git_dir, const char *path)
-{
-	struct strbuf buf = STRBUF_INIT;
-	int res;
-
-	if (strbuf_read_file(&buf, path, 0) < 0)
-		return error(_("could not read '%s'"), path);
-
-	res = stage(git_dir, &buf, path);
-
-	strbuf_release(&buf);
-	return res;
-}
-
-static int stage_directory(const char *git_dir, const char *path, int recurse)
-{
-	int at_root = !*path;
-	DIR *dir = opendir(at_root ? "." : path);
-	struct dirent *e;
-	struct strbuf buf = STRBUF_INIT;
-	size_t len;
-	int res = 0;
-
-	if (!dir)
-		return error(_("could not open directory '%s'"), path);
-
-	if (!at_root)
-		strbuf_addf(&buf, "%s/", path);
-	len = buf.len;
-
-	while (!res && (e = readdir(dir))) {
-		if (!strcmp(".", e->d_name) || !strcmp("..", e->d_name))
-			continue;
-
-		strbuf_setlen(&buf, len);
-		strbuf_addstr(&buf, e->d_name);
-
-		if ((e->d_type == DT_REG && stage_file(git_dir, buf.buf)) ||
-		    (e->d_type == DT_DIR && recurse &&
-		     stage_directory(git_dir, buf.buf, recurse)))
-			res = -1;
-	}
-
-	closedir(dir);
-	strbuf_release(&buf);
-	return res;
-}
+		goto cleanup;
+	}
 
-static int index_to_zip(const char *git_dir)
-{
-	struct child_process cp = CHILD_PROCESS_INIT;
-	struct strbuf oid = STRBUF_INIT;
+	if (cache_server_url ||
+	    supports_gvfs_protocol(dir, url, &cache_server_url)) {
+		if (set_config(config_path, "core.useGVFSHelper=true") ||
+		    set_config(config_path, "core.gvfs=150")) {
+			res = error(_("could not turn on GVFS helper"));
+			goto cleanup;
+		}
+		if (cache_server_url &&
+		    set_config(config_path,
+			       "gvfs.cache-server=%s", cache_server_url)) {
+			res = error(_("could not configure cache server"));
+			goto cleanup;
+		}
+	} else {
+		if (set_config(config_path, "core.useGVFSHelper=false") ||
+		    set_config(config_path, "remote.origin.promisor=true") ||
+		    set_config(config_path,
+			       "remote.origin.partialCloneFilter=blob:none")) {
+			res = error(_("could not configure partial clone in "
+				      "'%s'"), dir);
+			goto cleanup;
+		}
+	}
 
-	cp.git_cmd = 1;
-	strvec_pushl(&cp.args, "--git-dir", git_dir, "write-tree", NULL);
-	if (pipe_command(&cp, NULL, 0, &oid, the_hash_algo->hexsz + 1,
-			 NULL, 0))
-		return error(_("could not write temporary tree object"));
+	if (!full_clone &&
+	    (res = run_git(dir, "sparse-checkout", "init", "--cone", NULL)))
+		goto cleanup;
 
-	strbuf_rtrim(&oid);
-	child_process_init(&cp);
-	cp.git_cmd = 1;
-	strvec_pushl(&cp.args, "--git-dir", git_dir, "archive", "-o", NULL);
-	strvec_pushf(&cp.args, "%s.zip", git_dir);
-	strvec_pushl(&cp.args, oid.buf, "--", NULL);
-	strbuf_release(&oid);
-	return run_command(&cp);
-}
+	if (set_recommended_config(config_path))
+		return error(_("could not configure '%s'"), dir);
 
-#ifndef WIN32
-#include <sys/statvfs.h>
-#endif
+	/*
+	 * TODO: should we pipe the output and grep for "filtering not
+	 * recognized by server", and suppress the error output in
+	 * that case?
+	 */
```

Comment:
This can be polish for later, and I think we'd want to translate that error into something like "Remote does not support partial clone"